unlimited history

jaspermayone opened the pull request:
this should implement unlimited history. It did work locally for me, but i'd be curious for other people's thoughts / testing on implementation


closes #310

jaspermayone committed:
unlimited history

jaspermayone commented:
@p0deje lmk your thoughts on this. I know you were possibly waiting til some things changed, but I figured I would take a stab at it now.

weisJ commented on Maccy/Observables/History.swift:
```diff
@@ -70,6 +70,20 @@ class History { // swiftlint:disable:this type_body_length
+  @ObservationIgnored
+  var hasMoreItems: Bool {
+    all.count < totalCount
+  }
```
Maybe the pagination logic could be encapsulated in a separate class

jaspermayone replied:
closed in 8d02b0b525b561f54f0a0ea0630e8f0566212f48

weisJ commented on Maccy/Observables/History.swift:
```diff
@@ -122,6 +153,38 @@ class History { // swiftlint:disable:this type_body_length
+  private func loadPage() async throws {
+    let offset = currentPage * pageSize
+    var descriptor = FetchDescriptor<HistoryItem>(
+      sortBy: [SortDescriptor(\.lastCopiedAt, order: .reverse)]
```
This should take the current sorting mode into account. We need to make sure that all items on the next page are after the last item of the current page in sorting order.

jaspermayone replied:
closed in 8d02b0b525b561f54f0a0ea0630e8f0566212f48

weisJ commented on Maccy/Settings/StorageSettingsPane.swift:
```diff
@@ -93,12 +95,35 @@ struct StorageSettingsPane: View {
+          isOn: Binding(
+            get: { isUnlimitedHistory },
+            set: { newValue in
+              if newValue && size > 5000 {
```
Is `5000` deliberately the same as `historySizeThreshold`? If yes, it should be a constant referenced at both places.

jaspermayone replied:
closed in 8d02b0b525b561f54f0a0ea0630e8f0566212f48

weisJ commented on Maccy/Search.swift:
```diff
@@ -35,13 +35,23 @@ class Search {
+    var effectiveSearchMode = Defaults[.searchMode]
+
+    // Auto-disable fuzzy search for large histories
+    if Defaults[.isUnlimitedHistory] && History.shared.totalCount > historySizeThreshold {
+      if effectiveSearchMode == .fuzzy || effectiveSearchMode == .mixed {
+        effectiveSearchMode = .exact
+      }
+    }
```
This should be shown as a warning in the settings for the search mode.

jaspermayone replied:
closed in 8d02b0b525b561f54f0a0ea0630e8f0566212f48

weisJ commented on Maccy/Observables/History.swift:
```diff
@@ -122,6 +153,38 @@ class History { // swiftlint:disable:this type_body_length
+
+    let results = try Storage.shared.context.fetch(descriptor)
+    let newItems = sorter.sort(results).map { HistoryItemDecorator($0) }
+    all.append(contentsOf: newItems)
```
Previous pages should be unloaded, otherwise after a long usage we might be in the situation where the whole history is in-memory anyway.

jaspermayone replied:
closed in 8d02b0b525b561f54f0a0ea0630e8f0566212f48

weisJ reviewed (commented):
Mostly looking good to me, just a few comments. Currently the next page is fetched at the end of the current with a page size of 200. How about the following:

- Store (up to) 3 pages at a size of `100` each. The previous, the current and the next page.
- When reaching the end of the current page, load the page after the next and swap to the next page i.e.
	````
	previous = current
	current = next
	next = fetchNextPage()
	````
	(Similar scenario when scrolling up).
- Because the first page (and maybe last page) are frequently accessed (and easily jumped to with `CMD + UP/DOWN`, they could get special treatment and stored separately to avoid loading times.
- If a new item is added we need to make sure to remove it from any previous page it may be contained in and append it to the first (newest) page. Once the first page reaches 200 (or twice whatever the base page size is) it is split into two pages to prevent unlimited page size growing. Similar rebalancing needs to happen if the page the item is removed from grows too small (just discard the affected pages and reload them maybe).

I think this way we can almost certainly eliminate loading times when scrolling.
	In particular it does not keep the previous pages unnecessarily in memory when scrolling.

jaspermayone commented:
Will address!

jaspermayone committed:
address feedback

jaspermayone commented:
Should be good for re-review @weisJ!

jaspermayone committed:
Merge branch 'master' into feat/unlimited-history

jaspermayone committed:
fix copilot issues

weisJ commented:
Some more things I noticed:

When changing the settings for unlimited history the storage needs to recomputed to accommodate the new value.
Currently `loadPreviousItems()` is not called at all, so scrolling up does not work. In general SwifUI is a bit confused
when scroll ing because the list is (almost) always at the constant size of `pageSize`.
If you scroll to the end the content gets replaced but the scroll position is still "at the end of the list", which causes the pagination mechanism to immediately load all of the history and fast forward to the last page.

This wasn't a problem with the previous version, as there the list size increased linearly with more pages being loaded into memory. We probably need a custom scrollable list implementation. I think something as described here would be sufficient: https://nilcoalescing.com/blog/CustomLazyListInSwiftUI/

It assumes that every item has the same height, which is almost true for Maccy. There are two different heights, one for text and one for images. I cannot say how much more complicated that makes the implementation. Maybe precomputing the number of image entries per page could help...

The benefit with the custom scroll view approach would be that the scrollbar always reflects the actual size of the history list. Not just what is currently in memory.

weisJ commented on Maccy/Observables/PaginationManager.swift:
```diff
@@ -0,0 +1,325 @@
+    // If we're on the first page, also update currentPage
+    if currentPageIndex == 0 {
+      currentPage.insert(decorator, at: 0)
+    }
```
This could be avoided by storing pages not as lists by classes with a list inside. The current and first page could then just reference the same object.

Maybe this is also an opportunity to make `Page` adhere to the sequence protocol to avoid creating the concatenated array in `allLoadedItems`.

weisJ commented on Maccy/Observables/PaginationManager.swift:
```diff
@@ -0,0 +1,325 @@
+    nextPage.removeAll { $0.item == decorator.item }
+
+    // Insert at beginning of first page cache
+    firstPageCache.insert(decorator, at: 0)
```
This assumes that new items always are placed at the beginning. This is not true for e.g. "sort by time first copy" or "sort by number of copies".

weisJ commented on Maccy/Observables/PaginationManager.swift:
```diff
@@ -0,0 +1,325 @@
+    var descriptor = FetchDescriptor<HistoryItem>(
+      sortBy: [sortDescriptorForCurrentMode()]
+    )
+    descriptor.fetchLimit = pageSize
```
This currently assumes that all pages have the same size. This is not the case e.g. if something gets deleted or inserted. I have to think about that a good solution would be here. Maybe just invalidate the pages and refetch previous, current and next?

weisJ commented on Maccy/Observables/PaginationManager.swift:
```diff
@@ -0,0 +1,325 @@
+
+  /// Handle new item being added to history
+  @MainActor
+  func handleNewItem(_ decorator: HistoryItemDecorator) {
```
You handle insertion, but not yet deletion (except for `clear` and `clearAll`).

jaspermayone commented:
will implement latest rounds of feedback.

jaspermayone committed:
implement latest feedback

weisJ commented:
Sorry for taking so long to look into the new changes. My time was taken up by #1296. I will have a look later this week.

jaspermayone commented:
All good, 1296 looks super cool!

jaspermayone commented:
Following up @weisJ :)

weisJ commented:
I am sorry I still didn‘t get to fully go over the changes, so here are my preliminary impressions.

I tried out the code and manually set the page size to something small (10). The history list kept jumping while scrolling which is rather undesirable. Ultimately changing the underlying content of a list while scrolling will have to make use of a custom layout algorithm as described in https://nilcoalescing.com/blog/CustomLazyListInSwiftUI/. Otherwise we will not be able to get rid of the jumping.

jaspermayone commented:
> I am sorry I still didn‘t get to fully go over the changes, so here are my preliminary impressions.
> 
> I tried out the code and manually set the page size to something small (10). The history list kept jumping while scrolling which is rather undesirable. Ultimately changing the underlying content of a list while scrolling will have to make use of a custom layout algorithm as described in [nilcoalescing.com/blog/CustomLazyListInSwiftUI](https://nilcoalescing.com/blog/CustomLazyListInSwiftUI/). Otherwise we will not be able to get rid of the jumping.

All good @weisJ! I'll take a poke at this over the weekend.

jaspermayone committed:
Merge remote-tracking branch 'origin/master' into feat/unlimited-history

* origin/master: (65 commits)
  Fix title glitch over Maccy logo
  Cleanup unnecessary swiftlint comments
  Tidy up rounding window size including preview
  Highlight next found item after deletion
  Fix bug when only 1 item cannot be delete
  Reset search after deleting item
  Reset search after toggling pin
  Fix window height with small number of items
  Fix swiftlint error
  Fix missing unpin key translations
  Fix invalid key in preview tooltip
  Reset preview state when re-opening popup
  Round slideout to avoid 1-2 px movement to the left
  Fix auto-opening
  Fix incorrect hr/uz translations
  Fix wrong tooltip on delete button
  Move toolbar to the preview view
  Switch to Control-Space to open preview
  Add translations for slideout preview
  Fix swiftlint errors
  ...

jaspermayone committed:
add new layout

jaspermayone commented:
@weisJ this should be a bit better!

jaspermayone commented:
Hey @weisJ, just bumping this :)

weisJ commented on Maccy/Views/VirtualizedLayout.swift:
```diff
@@ -0,0 +1,191 @@
+    // Total height is fixed based on total item count - never changes during scrolling
+    let height = CGFloat(totalItemCount) * itemHeight
+    return CGSize(width: width, height: height)
+  }
```
This needs to take into account the number of items with an image, as they have a different height from normal items.

jaspermayone replied:
fixing!

jaspermayone committed:
fix: account for image item heights in virtualized layout

Track sorted indices of image-bearing items in PaginationManager and use
them in VirtualizedLayout to compute exact total height and per-row
y-offsets, instead of assuming a uniform item height.

jaspermayone commented:
@weisJ this is good for re-review whenever you have time :)

weisJ commented:
Thanks for the bumb. Sadly I have been swamped lately. Trying to find some time in the next few days though :)

jaspermayone commented:
Hey @p0deje or @weisJ! Checking in :)

jaspermayone committed:
refactor: map scroll position to pages, fix row heights, make pagination testable

Replace the sentinel-based loading with a direct mapping from scroll
offset to row range to page indices, so fast scrolling can no longer
miss loads. PaginationManager is now generic over a PaginatedItemSource,
allowing tests to drive it with a synthetic source (large list of
numbers, small pages). Row geometry lives in VirtualListMetrics, and
image rows are indexed in the same (unpinned-only) space the list
renders, fixing the mismatched item heights.

jaspermayone commented on Maccy/Observables/PaginationManager.swift:
```diff
@@ -0,0 +1,393 @@
+      indices.append(index)
+    }
+    imageItemIndices = indices
+  }
```
Reworked: image presence is now determined from `HistoryItemContent` types via a predicate (no image data is loaded or decoded), and the index pass over sorted items touches metadata only — no in-memory re-sort. Still O(n) per mutation, but that cost is inherent to needing global row indices for the layout.

jaspermayone commented on Maccy/Observables/History.swift:
```diff
@@ -99,16 +141,33 @@ class History: ItemsContainer { // swiftlint:disable:this type_body_length
   @MainActor
   func load() async throws {
-    let descriptor = FetchDescriptor<HistoryItem>()
-    let results = try Storage.shared.context.fetch(descriptor)
-    all = sorter.sort(results).map { HistoryItemDecorator($0) }
-    items = all
+    all.removeAll()
 
-    limitHistorySize(to: Defaults[.size])
+    if Defaults[.isUnlimitedHistory] {
+      // Use pagination manager for unlimited history
+      try await paginationManager.load()
+      all = paginationManager.allLoadedItems
+    } else {
+      // Load all items for limited history
+      let descriptor = FetchDescriptor<HistoryItem>()
+      let results = try Storage.shared.context.fetch(descriptor)
+      all = sorter.sort(results).map { HistoryItemDecorator($0) }
+      limitHistorySize(to: Defaults[.size])
+    }
+
+    items = all
```
Partially addressed: while a search is active the popup now falls back to the standard (in-memory) list, so results render correctly — but search still only matches items in the loaded window. Searching storage directly in unlimited mode is a real gap; I think it deserves its own follow-up rather than growing this PR further.

jaspermayone commented on Maccy/Observables/History.swift:
```diff
@@ -117,6 +176,66 @@ class History: ItemsContainer { // swiftlint:disable:this type_body_length
+    do {
+      try await paginationManager.loadPreviousWindow()
+      all = paginationManager.allLoadedItems
+      items = all
+      updateUnpinnedShortcuts()
+      AppState.shared.popup.needsResize = true
+    } catch {
+      logger.error("Failed to load previous items: \(error.localizedDescription)")
+    }
+  }
+
+  @MainActor
+  func jumpToFirst() async {
+    guard Defaults[.isUnlimitedHistory] else { return }
+
+    do {
+      try await paginationManager.jumpToFirst()
+      all = paginationManager.allLoadedItems
+      items = all
+      updateUnpinnedShortcuts()
+      AppState.shared.popup.needsResize = true
+    } catch {
+      logger.error("Failed to jump to first: \(error.localizedDescription)")
+    }
+  }
+
+  @MainActor
+  func jumpToLast() async {
+    guard Defaults[.isUnlimitedHistory] else { return }
+
+    do {
+      try await paginationManager.jumpToLast()
+      all = paginationManager.allLoadedItems
+      items = all
+      updateUnpinnedShortcuts()
+      AppState.shared.popup.needsResize = true
+    } catch {
+      logger.error("Failed to jump to last: \(error.localizedDescription)")
+    }
+  }
```
The pagination logic is now generic over a `PaginatedItemSource` and covered by `PaginationManagerTests` (synthetic integer source, page size 20) plus `VirtualListMetricsTests` for the row geometry. I did not add `HistoryTests` cases because that suite is skipped in `Maccy.xctestplan`, so they would not run.
